pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -94,28 +98,365 @@ func (p *staticPolicy) Start(logger klog.Logger, s state.State) error {
+				}
+				for resourceName, poolSize := range poolForStandardInits {
+					if containerReqs[resourceName] > poolSize {
+						return fmt.Errorf("pod rejected, largest exclusive init container requests %d of %s, but only %d available", containerReqs[resourceName], resourceName, poolSize)
+					}
+				}
+			} else {
+				// This check ensures that if there are any standard init containers that need a
+				// shared pool, that pool is not empty. An empty pool would occur if the pod
+				// budget is fully consumed by sidecars.
+				for resourceName, poolSize := range podTotalMemory {
+					if sumOfLongRunningExclusiveMemory[resourceName] >= poolSize {
+						return fmt.Errorf("pod rejected, pod has shared init containers but no %s available for them", resourceName)
+					}
+				}
+			}
+
+			continue
+		}
+
+		if len(containerReqs) == 0 {
+			hasSharedAppContainers = true
+			continue
+		}
+
+		for resourceName, reqAmount := range containerReqs {
+			sumOfLongRunningExclusiveMemory[resourceName] += reqAmount
+		}
+	}
+
+	// Check for empty shared pool for concurrently running containers (app and restartable init containers).
+	for _, container := range pod.Spec.Containers {
+		containerReqs, err := getContainerRequestedResources(logger, pod, &container)
+		if err != nil {
+			return err
+		}
+
+		if len(containerReqs) == 0 {
+			hasSharedAppContainers = true
+			continue
+		}
+
+		for resourceName, val := range containerReqs {
+			sumOfLongRunningExclusiveMemory[resourceName] += val
+		}
+	}
+
+	// This check ensures that if there are any standard containers that need a shared
+	// pool, that pool is not empty. An empty pool would occur if the pod budget is
+	// fully consumed by sidecars and main containers with exclusive resources.
+	if hasSharedAppContainers {
+		for resourceName, totalPodResource := range podTotalMemory {
+			if sumOfLongRunningExclusiveMemory[resourceName] >= totalPodResource {
+				return fmt.Errorf("pod rejected, sum of exclusive container %s requests equals pod budget, leaving no memory for shared containers", resourceName)
+			}
+		}
+	}
+
+	return nil
+}
+
+// AllocatePod is the entry point for pod-level resource allocation.
+// It's called once per pod by the Topology Manager's pod-scope admit handler.
+// The logic here allocates a single NUMA-aligned "bubble" of memory for the
+// entire pod. All containers within the pod will share this NUMA binding.
+func (p *staticPolicy) AllocatePod(logger klog.Logger, s state.State, pod *v1.Pod, hint topologymanager.TopologyHint) (rerr error) {
+	logger = klog.LoggerWithValues(logger, "pod", klog.KObj(pod))
+	logger.V(4).Info("AllocatePod called for pod-level managed pod", "hint", hint.NUMANodeAffinity)
+
+	qos := v1qos.GetPodQOS(pod)
+	if qos != v1.PodQOSGuaranteed {
+		return nil
+	}
+
+	// 1. Calculate memory resources.
+	podTotalMemory, err := getPodRequestedResources(logger, pod)
```

Comment:
can `podTotalMemory` ever be empty or nil, likewise we had zero guaranteed CPUs?